Sources/System/MachPort.swift

```diff
@@ -0,0 +1,318 @@
+    ///
+    /// Upon destruction a send-once notification will be sent to the
+    /// receiving end.
+    struct SendOnceRight : MachPortRight {}
+
+    /// Create a connected pair of rights, one receive, and one send.
+    ///
+    /// This function will abort if the rights could not be created.
+    /// Callers may assert that valid rights are always returned.
+    static func allocatePortRightPair() -> (Mach.Port<Mach.ReceiveRight>, Mach.Port<Mach.SendRight>) {
+        var name = mach_port_name_t(MACH_PORT_NULL)
+        let secret = mach_port_context_t(arc4random())
+        withUnsafeMutablePointer(to: &name) { name in
+            var options = mach_port_options_t()
+            options.flags = UInt32(MPO_INSERT_SEND_RIGHT);
+            withUnsafeMutablePointer(to: &options) { options in
+                let kr = mach_port_construct(mach_task_self_, options, secret, name)
+                assert(kr == KERN_SUCCESS)
+            }
+        }
+        return (Mach.Port<Mach.ReceiveRight>(name: name, context: secret), Mach.Port<Mach.SendRight>(name: name))
+    }
+}
+
+extension Mach.Port where RightType == Mach.ReceiveRight {
+    /// Transfer ownership of an existing, unmanaged, but already guarded,
+    /// Mach port right into a Mach.Port by name.
+    ///
+    /// This initializer aborts if name is MACH_PORT_NULL.
+    ///
+    /// If the type of the right does not match the type T of Mach.Port<T>
+    /// being constructed, the behavior is undefined.
+    ///
+    /// The underlying port right will be automatically deallocated when
+    /// the Mach.Port object is destroyed.
+    init(name:mach_port_name_t, context:mach_port_context_t) {
+        self.name = name
+        self.context = context
+    }
+
+    /// Allocate a new Mach port with a receive right, creating a
+    /// Mach.Port<Mach.ReceiveRight> to manage it.
+    ///
+    /// This initializer will abort if the right could not be created.
+    /// Callers may assert that a valid right is always returned.
+    init() {
+        var storage:mach_port_name_t = 0
+        withUnsafeMutablePointer(to:&storage) { storage in
+            let kr = mach_port_allocate(mach_task_self_, MACH_PORT_RIGHT_RECEIVE, storage)
+            assert(kr == KERN_SUCCESS)
+        }
+
+        // name-only init will guard ReceiveRights
+        self.init(name:storage)
+    }
+
+
+    /// Transfer ownership of the underlying port right to the caller.
+    ///
+    /// Returns a tuple containing the Mach port name representing the right,
+    /// and the context value used to guard the right.
+    ///
+    /// This operation liberates the right from management by the Mach.Port,
+    /// and the underlying right will no longer be automatically deallocated.
+    ///
+    /// After this function completes, the Mach.Port is destroyed and no longer
+    /// usable.
+    __consuming func relinquish() -> (mach_port_name_t, mach_port_context_t) {
```

Comment:
At some point we'd talked about using labels here to improve the clarity of the return value:
```swift
__consuming func relinquish() -> (name: mach_port_name_t, `guard`: mach_port_context_t)
```

Would allow users to use
```swift
value.name
value.guard
```